[WIP] refactor module discovery

eoneill opened the pull request:
This is a pretty substantial change, but it should address some weird edge cases we've seen along the way with module discovery (e.g. #70).

The current approach attempts to discover modules on every `@import` statement. This is both costly and non-deterministic.

The approach proposed here is to discover modules once (when `new Eyeglass()` instance is created). This collection of modules is the source of truth for the life-span of the instance (stored on `this.modules`). We use this for all module lookups and we preserve the tree hierarchy to determine scope of a module.

This also allows us to expose a tree view of the modules for easy debugging (per #66). For example, here's the result of `this.modules.getGraph()` on the `function_modules` test case:
```
:root(function_modules)
├─┬ fn_module_a
│ └── transitive_fn
├── fn_module_b
└── eyeglass
```

Please provide any feedback and I'll work on polishing this up a bit and ensuring all the test are passing again.

Jakobo commented on lib/util/eyeglass_modules.js:
```diff
@@ -0,0 +1,283 @@
+var uuid = require("uuid");
+
+var EYEGLASS_KEYWORD = "eyeglass-module";
+var ROOT_ID = ":root-" + uuid.v4();
```
reasons for uuid v reserved name or just non-npm compatible package name?

eoneill replied:
I was going somewhere with that, but I backtracked and probably don't need it anymore.

Jakobo commented on lib/util/eyeglass_modules.js:
```diff
@@ -0,0 +1,283 @@
+  return registry;
+}
+
+function checkVersions(versions, latest) {
```
`checkVersions` seems to return incompatible versions. Either the fn name or the return object should probably change.

eoneill replied:
Yeah, this definitely needs a test. I expect to be adding several tests as part of this.

eoneill commented:
Part of me wants to move `EyeglassModules` into a separate repo with a dedicated test suite (e.g. `eyeglass-core-modules`)

Jakobo commented on lib/util/eyeglass_modules.js:
```diff
@@ -0,0 +1,283 @@
+  });
+}
+
+function dedupeModules(modules) {
```
This operates in place and returns an error array. Will definitely want to make sure that's noted and has a test

Jakobo commented on lib/util/eyeglass_modules.js:
```diff
@@ -0,0 +1,283 @@
+  return require("archy")(hierarchy);
+};
+
+function getDecoratedRootName() {
```
Semantics. Is the project moving towards `.call` to create pseudo private scope?

eoneill replied:
I like it, but not entirely opposed to exposing things on the prototype.

Jakobo commented on lib/util/eyeglass_modules.js:
```diff
@@ -0,0 +1,283 @@
+  return Object.keys(dependencies).length ? dependencies : null;
+}
+
+function tick(str) {
```
Is this the only place we're bothering with backticks around path stuff or are we missing a util file as part of the refactor?

eoneill replied:
I was using it in more places, but then backtracked on the backticks. We should be consistent throughout eyeglass warnings / errors.

Jakobo commented on lib/util/package.js:
```diff
@@ -0,0 +1,50 @@
+"use strict";
```
6 months after the initial discovery code was written, this is worthy of 100 <3s. Thank you.

Jakobo commented on package.json:
```diff
@@ -28,25 +28,29 @@
-    "mocha": "^2.2.5",
-    "should": ">= 0.0.1"
+    "mocha": "^2.3.3",
+    "should": ">=7.1.0"
```
Is leaving `>=` intentional? Can we use `^7.1.0` instead?

eoneill replied:
I should definitely be using `^7.1.0`. Will update.

chriseppstein commented:
> Part of me wants to move `EyeglassModules` into a separate repo with a dedicated test suite (e.g. `eyeglass-core-modules`)

I'd rather we didn't.

chriseppstein commented:
I had set up the travis build so that coverage was only ran once instead of with every build permutation and it looks like you've removed that. :-1:

eoneill commented:
> I had set up the travis build so that coverage was only ran once instead of with every build permutation and it looks like you've removed that. :-1:

I'll revert this change.

chriseppstein commented on lib/util/eyeglass_modules.js:
```diff
@@ -0,0 +1,283 @@
+    var modulePath = path.dirname(pkg.path);
+    if (typeof pkg.data.version === "object") {
+      console.log("pkg.data.version", pkg.data.version);
+    }
```
remove debugging code?

chriseppstein commented on lib/util/eyeglass_modules.js:
```diff
@@ -0,0 +1,283 @@
+      path: modulePath,
+      version: pkg.data.version,
+      main: getModuleExports(pkg.data, modulePath),
+      isEyeglassModule: isEyeglassMod,
```
why would `isEyeglassModule` ever be false?

eoneill replied:
If the project root itself is not an eyeglass module. That's the only case.

This is important so that we can correctly preserve the dependency tree for checking access permission.

chriseppstein commented on lib/util/eyeglass_modules.js:
```diff
@@ -0,0 +1,283 @@
+  if (branch) {
+    // if the branch itself is a module, add it...
+    if (branch.isEyeglassModule) {
+      registry = registry || {};
```
better to set this default at the start of the function instead of repeating it several times.

chriseppstein commented on lib/util/eyeglass_modules.js:
```diff
@@ -0,0 +1,283 @@
+    var dependencies = branch.dependencies;
+
+    if (dependencies) {
+      var result = Object.keys(dependencies).reduce(function(registry, name) {
```
pretty sure using the same argument name in a nested function is going to be a linter error -- are you running the linter in development?

eoneill replied:
Yes I am. It doesn't complain, but I was going to change this anyway.

chriseppstein commented:
When running the tests they all pass but there's some console logging:

```
{ [Error: Cannot find module 'module_a/package.json'] code: 'MODULE_NOT_FOUND' } [ 'Error: Cannot find module \'module_a/package.json\'',
  '    at Function.Module._resolveFilename (module.js:337:15)',
  '    at resolve (/Users/ceppstei/Projects/eyeglass/lib/util/resolve.js:9:238)',
  '    at /Users/ceppstei/Projects/eyeglass/lib/util/eyeglass_modules.js:9:14880',
  '    at Array.reduce (native)',
  '    at discoverModules (/Users/ceppstei/Projects/eyeglass/lib/util/eyeglass_modules.js:9:14749)',
  '    at resolveModule (/Users/ceppstei/Projects/eyeglass/lib/util/eyeglass_modules.js:9:13389)',
  '    at new EyeglassModules (/Users/ceppstei/Projects/eyeglass/lib/util/eyeglass_modules.js:9:929)',
  '    at Object.Eyeglass (/Users/ceppstei/Projects/eyeglass/lib/index.js:9:1343)',
  '    at decorator (/Users/ceppstei/Projects/eyeglass/lib/index.js:9:4313)',
  '    at Object.module.exports.sassOptions (/Users/ceppstei/Projects/eyeglass/test/testutils.js:9:3084)' ]
```

This occurs on the test `errors on invalid module`. Is this output expected? If so, we should capture stderr/out and assert that it's output. There are test utilities for capturing output if you didn't see them yet.

eoneill commented:
> When running the tests they all pass but there's some console logging:

I need to clean this up still. Tests are not all good yet and I have more tests to write. At this point, I'm mainly looking for preliminary feedback that this is an overall sound approach. There's a lot of cleanup to come, but I don't want to jump down a rabbit hole if the idea is off base.

chriseppstein commented:
>  I don't want to jump down a rabbit hole if the idea is off base.
 
I don't see any theoretical issues with it. I was memoizing module information already so it's not clear how big of a performance gain this will actually provide, but it does seem to be a nicer abstraction to work with in code.

eoneill commented:
All tests are now passing with `node-sass@3.3.3`, unfortunately `node-sass@3.4.0` causes tests to fail `@HEAD`, so please run tests locally to verify.

I think I addressed all the comments above, but let me know if I missed anything.

eoneill commented:
Barring node-sass issues, this should be good for review/merge now. I added code docs as well as some additional tests for `EyeglassModules`

Thanks for your patience.

chriseppstein commented on docs/TROUBLESHOOTING.md:
```diff
@@ -8,3 +8,14 @@ This happens during your `npm install` process, and in our case is almost certai
+
+# Logging debug messages
+
+If you're still running into issues, you can gain some insight by enabling some debug flags. To enable this logging, set an environment variable named `DEBUG` to one or more of the following:
```
Let's link to the [debug module here](https://www.npmjs.com/package/debug).
